File: python_solutions/1275.find_winner_on_a_tictactoe_game.py

# https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/description/
# Easy


class Solution:
    def tictactoe(self, moves):
        if len(moves) < 5:
            return 'Pending'

        grids = [[''] * 3 for _ in range(3)]
        for i, m in enumerate(moves):
            grids[m[0]][m[1]] = 'O' if i & 1 else 'X'

        last_sys = 'X' if len(moves) & 1 else 'O'
        winner = 'A' if last_sys == 'X' else 'B'

        for i in range(3):
            if ''.join(grids[i]) == last_sys * \
                    3 or ''.join([grids[0][i], grids[1][i], grids[2][i]]) == last_sys * 3:
                return winner

        
        if ''.join([grids[0][0], grids[1][1], grids[2][2]]) == last_sys * \
                3 or ''.join([grids[2][0], grids[1][1], grids[0][2]]) == last_sys * 3:
            return winner
        
        return 'Draw' if len(moves) == 9 else 'Pending'



s = Solution()
moves = [[0, 0], [1, 1], [0, 1], [0, 2], [1, 0], [2, 0]]
moves = [[0,0],[2,0],[1,1],[2,1],[2,2]]
moves = [[0,0],[1,1],[2,0],[1,0],[1,2],[2,1],[0,1],[0,2],[2,2]]
moves = [[1,0],[2,2],[2,0],[0,1],[1,1]]
moves = [[2,0],[1,1],[0,2],[2,1],[1,2],[1,0],[0,0],[0,1]]
print(s.tictactoe(moves))

File: python_solutions/1275.find_winner_on_a_tictactoe_game.cpp

// #
// https://leetcode.com/problems/find-winner-on-a-tic-tac-toe-game/description/
// # Easy

class Solution {
public:
  bool is_win(vector<vector<char>> &gd, char c) {
    for (int i = 0; i < 3; i++) {
      if (gd[i][0] == c && gd[i][1] == c && gd[i][2] == c)
        return true;
      if (gd[0][i] == c && gd[1][i] == c && gd[2][i] == c)
        return true;
    }
    if (gd[0][0] == c && gd[1][1] == c && gd[2][2] == c)
      return true;
    if (gd[2][0] == c && gd[1][1] == c && gd[0][2] == c)
      return true;

    return false;
  }

  string tictactoe(vector<vector<int>> &ms) {
    if (ms.size() < 5)
      return "Pending";
    vector<vector<char>> gd(3, vector<char>(3, '#'));

    for (int i = 0; i < ms.size(); i++) {
      gd[ms[i][0]][ms[i][1]] = i & 1 ? 'O' : 'X';
    }

    char c = ms.size() & 1 ? 'X' : 'O';
    bool res = is_win(gd, c);
    if (res)
      return c == 'X' ? "A" : "B";
    else
      return ms.size() == 9 ? "Draw" : "Pending";
  }
};